compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/AnnotationDesugar.java

```diff
@@ -163,43 +165,126 @@ void rewritePackageAnnotations(BLangPackage pkgNode, SymbolEnv env) {
+                    owner);
             if (lambdaFunction != null) {
-                BType type = classDefinition.getBType();
+                BType type = classDef.getBType();
                 if (Symbols.isFlagOn(type.flags, Flags.OBJECT_CTOR)) {
-                    // Add the lambda/invocation in a temporary block.
-                    BLangBlockStmt target = (BLangBlockStmt) TreeBuilder.createBlockNode();
-                    BLangBlockFunctionBody initBody = (BLangBlockFunctionBody) initFunction.body;
-                    target.pos = initBody.pos;
-                    addLambdaToGlobalAnnotMap(classDefinition.name.value, lambdaFunction, target);
-                    int index = calculateIndex((initBody).stmts, type.tsymbol);
-                    for (BLangStatement stmt : target.stmts) {
-                        initBody.stmts.add(index++, stmt);
+                    if (normalMode) {
+                        // Add the lambda/invocation in a temporary block.
+                        BLangBlockStmt target = (BLangBlockStmt) TreeBuilder.createBlockNode();
+                        target.pos = initBody.pos;
+                        addLambdaToGlobalAnnotMap(classDef.name.value, lambdaFunction, target);
+                        int index = calculateIndex((initBody).stmts, type.tsymbol);
+                        for (BLangStatement stmt : target.stmts) {
+                            initBody.stmts.add(index++, stmt);
+                        }
+                    } else {
+                        // Add the lambda/invocation in a temporary block.
+                        LocationData locationData = new LocationData(pkgID, owner, pos, initBody);
+                        addAnnotationLambdaToGlobalAnnotationMapWithBlockDef(pkgNode, classDef, locationData,
+                                lambdaFunction);
                     }
                 } else {
-                    addInvocationToGlobalAnnotMap(classDefinition.name.value, lambdaFunction, initFunction.body);
+                    addInvocationToGlobalAnnotMap(classDef.name.value, lambdaFunction, initBody);
                 }
             }
         }
     }
 
+    private void addAnnotationLambdaToGlobalAnnotationMapWithBlockDef(BLangPackage pkgNode,
+                                                                      BLangClassDefinition classDef,
+                                                                      LocationData location,
+                                                                      BLangLambdaFunction lambdaFunction) {
+        BLangBlockStmt target = (BLangBlockStmt) TreeBuilder.createBlockNode();
+        PackageID pkgID = location.pkgID;
+        Location pos = location.pos;
+        BSymbol owner = location.owner;
+
+        BLangBlockFunctionBody initBody = location.body;
+        target.pos = pos;
+
+        String lamdaName = "$anon" + classDef.name.value + "_lambda_" + lambdaFunction.function.name.value;
```

Comment:
We can use `BLangAnonymousModelHelper` here? can be addressed in https://github.com/ballerina-platform/ballerina-lang/issues/34395